libcxx/include/__flat_map/flat_map.h

```diff
@@ -0,0 +1,1340 @@
+  _LIBCPP_HIDE_FROM_ABI const_reverse_iterator crend() const noexcept { return const_reverse_iterator(begin()); }
+
+  // [flat.map.capacity], capacity
+  [[nodiscard]] _LIBCPP_HIDE_FROM_ABI bool empty() const noexcept { return __containers_.keys.empty(); }
+
+  _LIBCPP_HIDE_FROM_ABI size_type size() const noexcept { return __containers_.keys.size(); }
+
+  _LIBCPP_HIDE_FROM_ABI size_type max_size() const noexcept {
+    return std::min<size_type>(__containers_.keys.max_size(), __containers_.values.max_size());
+  }
+
+  // [flat.map.access], element access
+  _LIBCPP_HIDE_FROM_ABI mapped_type& operator[](const key_type& __x)
+    requires is_constructible_v<mapped_type>
+  {
+    return try_emplace(__x).first->second;
+  }
+
+  _LIBCPP_HIDE_FROM_ABI mapped_type& operator[](key_type&& __x)
+    requires is_constructible_v<mapped_type>
+  {
+    return try_emplace(std::move(__x)).first->second;
+  }
+
+  template <class _Kp>
+    requires(__is_compare_transparent && is_constructible_v<key_type, _Kp> && is_constructible_v<mapped_type> &&
+             !is_convertible_v<_Kp &&, const_iterator> && !is_convertible_v<_Kp &&, iterator>)
+  _LIBCPP_HIDE_FROM_ABI mapped_type& operator[](_Kp&& __x) {
+    return try_emplace(std::forward<_Kp>(__x)).first->second;
+  }
+
+  _LIBCPP_HIDE_FROM_ABI mapped_type& at(const key_type& __x) {
+    auto __it = find(__x);
+    if (__it == end()) {
+      std::__throw_out_of_range("flat_map::at(const key_type&): Key does not exist");
+    }
+    return (*__it).second;
+  }
+
+  _LIBCPP_HIDE_FROM_ABI const mapped_type& at(const key_type& __x) const {
+    auto __it = find(__x);
+    if (__it == end()) {
+      std::__throw_out_of_range("flat_map::at(const key_type&) const: Key does not exist");
+    }
+    return (*__it).second;
+  }
+
+  template <class _Kp>
+    requires __is_compare_transparent
+  _LIBCPP_HIDE_FROM_ABI mapped_type& at(const _Kp& __x) {
+    static_assert(requires { find(__x); }, "flat_map::at(const K& x): find(x) needs to be well-formed");
+    auto __it = find(__x);
+    if (__it == end()) {
+      std::__throw_out_of_range("flat_map::at(const K&): Key does not exist");
+    }
+    return (*__it).second;
+  }
+
+  template <class _Kp>
+    requires __is_compare_transparent
+  _LIBCPP_HIDE_FROM_ABI const mapped_type& at(const _Kp& __x) const {
+    static_assert(requires { find(__x); }, "flat_map::at(const K& x) const: find(x) needs to be well-formed");
+    auto __it = find(__x);
+    if (__it == end()) {
+      std::__throw_out_of_range("flat_map::at(const K&) const: Key does not exist");
+    }
+    return (*__it).second;
+  }
+
+  // [flat.map.modifiers], modifiers
+  template <class... _Args>
+    requires is_constructible_v<pair<key_type, mapped_type>, _Args...>
+  // todo: LWG
+  // insufficiently constrained. key and values need to be move constructible
+  _LIBCPP_HIDE_FROM_ABI pair<iterator, bool> emplace(_Args&&... __args) {
```

Comment:
This needs and LWG issue

Reply:
# Title
`flat_map::emplace` is insufficiently constrained

# Contents:
Currently the member function
```
template<class... Args> pair<iterator, bool> emplace(Args&&... args);
```
has the following constraints

> Constraints: is_constructible_v<pair<key_type, mapped_type>, Args...> is true
(http://eel.is/c++draft/flat.map#modifiers-1.sentence-1)

However, the `emplace` function would construct a `pair` from the `args...`, then move the `first` and `second` into the `KeyContainer` and `MappedContainer` respectively. This would requires the `Key` and `Mapped` to be move constructible.

# Proposed Wording

```
template<class... Args> pair<iterator, bool> emplace(Args&&... args);
```

<del>Constraints: `is_constructible_v<pair<key_type, mapped_type>, Args...>` is `true`</del>
<ins>Constraints:
- `is_constructible_v<pair<key_type, mapped_type>, Args...>` is `true`
- `is_move_constructible_v<key_type>` is `true`
- `is_move_constructible_v<mapped_type>` is `true`</ins>

Reply:
In our implementation, let's add the missing constraints (i.e. assume that the LWG issue is going to get resolved the way we want it).

Reply:
I sent the email and Daniel replied that "we never impose is_move_constructible constraints on key or value types of containers or adaptors". Which means that the standard wants us to hard error, which is fair

Reply:
Ack, got it.